File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int ps[]=new int[heights.length];
		int ns[]=new int[heights.length];
		ns=NextSmaller(heights);
		ps=PreSmaller(heights);
		int max=0;
		for(int i=0;i<heights.length;i++) {
			int curr=heights[i]*(ns[i]-ps[i]-1);
			max=Math.max(max, curr);
		}
        return max;
    }
    static int[] NextSmaller(int arr[]){
     		int ns[]=new int[arr.length];
		Deque<Integer> stack = new LinkedList<>();
		for(int i=arr.length-1;i>=0;i--) {
			while(!stack.isEmpty() && arr[i]<=arr[stack.peek()])
				stack.pop();
			if(!stack.isEmpty()) {
				ns[i]=stack.peek();
			}
			else
				ns[i]=arr.length;
			stack.push(i);
		}
		return ns;   
    }
    static int[] PreSmaller(int arr[]){
     int ps[]=new int[arr.length];
		Deque<Integer> stack = new LinkedList<>();
		for(int i=0;i<arr.length;i++) 
		{
			while(!stack.isEmpty() && arr[stack.peek()]>=arr[i])
				stack.pop();
			if(!stack.isEmpty()) {
				ps[i]=stack.peek();
			}
			else {
				ps[i]=-1;
			}
			stack.push(i);
		}
		return ps;   
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.kt
import java.util.ArrayDeque

class Solution {
    fun largestRectangleArea(heights: IntArray): Int {
        // have to traverse to left and right until we find the bar whose 
        // height is less than the current height and calculate the area
        
        var left=IntArray(heights.size)
        var right=IntArray(heights.size)
        val lstack=ArrayDeque<Int>()
        val rstack=ArrayDeque<Int>()
        var area=0
        
        for(i in 0 until heights.size){
            
            while(lstack.isNotEmpty() && heights[i]<=heights[lstack.peek()]){
                lstack.pop()
            }
            
            if(lstack.isEmpty()){
                left[i]=0
                
            }else{
                left[i]=lstack.peek()+1
            }
            lstack.push(i) 
        }
        
         for(i in heights.size-1 downTo 0){
            
            while(rstack.isNotEmpty() && heights[i]<=heights[rstack.peek()]){
                rstack.pop()
            }
            
            if(rstack.isEmpty()){
                right[i]=heights.size-1
                
            }else{
                right[i]=rstack.peek()-1
            }
            rstack.push(i)
            if(area<(right[i]-left[i]+1)*heights[i]){
                area=(right[i]-left[i]+1)*heights[i]
            }
        }
        
        
        return area
    }
}